File: fastApiProject/service/redisService.py
import os
import json
import uuid
from typing import Optional, Dict, Any
import redis
from datetime import datetime, timedelta


class RedisService:
    def __init__(self):
        self.host = os.getenv("REDIS_HOST", "localhost")
        self.port = int(os.getenv("REDIS_PORT", 6379))
        self.db = int(os.getenv("REDIS_DB", 0))
        self.password = os.getenv("REDIS_PASSWORD", None)
        self.session_ttl = int(os.getenv("REDIS_SESSION_TTL", 86400))

        self.client = None
        self._connect()

    def _connect(self):
        """连接Redis服务器"""
        try:
            self.client = redis.Redis(
                host=self.host,
                port=self.port,
                db=self.db,
                password=self.password if self.password else None,
                decode_responses=True,  # 自动解码为字符串
                socket_connect_timeout=5,
                socket_timeout=5,
                retry_on_timeout=True
            )
            # 测试连接
            self.client.ping()
        except Exception as e:
            self.client = None

    def is_connected(self) -> bool:
        """检查Redis连接状态"""
        if not self.client:
            return False
        try:
            self.client.ping()
            return True
        except:
            return False

    def create_session(self, user_id: int, user_data: Dict[str, Any]) -> Optional[str]:
        """创建用户会话"""
        if not self.is_connected():
            return None

        try:
            session_id = str(uuid.uuid4())
            session_key = f"session:{session_id}"
            user_key = f"user_session:{user_id}"

            # 存储会话数据
            session_data = {
                "user_id": user_id,
                "created_at": datetime.now().isoformat(),
                **user_data
            }

            # 设置会话
            self.client.setex(
                session_key,
                self.session_ttl,
                json.dumps(session_data)
            )

            # 关联用户ID和会话ID
            self.client.sadd(user_key, session_id)
            self.client.expire(user_key, self.session_ttl)

            return session_id
        except Exception:
            return None

    def get_session(self, session_id: str) -> Optional[Dict[str, Any]]:
        """获取会话数据"""
        if not self.is_connected():
            return None

        try:
            session_key = f"session:{session_id}"
            data = self.client.get(session_key)
            if data:
                # 更新会话过期时间
                self.client.expire(session_key, self.session_ttl)
                return json.loads(data)
            return None
        except Exception as e:
            print(f"获取会话失败: {e}")
            return None

    def delete_session(self, session_id: str) -> bool:
        """删除会话"""
        if not self.is_connected():
            return False

        try:
            session_key = f"session:{session_id}"
            data = self.client.get(session_key)
            if data:
                session_data = json.loads(data)
                user_id = session_data.get("user_id")

                # 删除会话
                self.client.delete(session_key)

                # 从用户会话集合中移除
                if user_id:
                    user_key = f"user_session:{user_id}"
                    self.client.srem(user_key, session_id)

            return True
        except Exception as e:
            print(f"删除会话失败: {e}")
            return False

    def delete_user_sessions(self, user_id: int) -> bool:
        """删除用户的所有会话"""
        if not self.is_connected():
            return False

        try:
            user_key = f"user_session:{user_id}"
            session_ids = self.client.smembers(user_key)

            # 删除所有会话
            for session_id in session_ids:
                session_key = f"session:{session_id}"
                self.client.delete(session_key)

            # 删除用户会话集合
            self.client.delete(user_key)
            return True
        except Exception as e:
            print(f"删除用户会话失败: {e}")
            return False

    def check(self, token: str = None) -> bool:
        # 检查token是否存在
        if not self.client:
            return False
        return self.client.exists(f"token:{token}")
    
    def check_set_member(self, set_name: str, member: str) -> bool:
        # 检查集合中是否包含指定成员（处理带引号的情况）
        if not self.client:
            return False
        # 检查原始值
        if self.client.sismember(set_name, member):
            return True
        # 检查带双引号的值
        quoted_member = f'"{member}"'
        quoted_set_name = f'"{set_name}"'
        return self.client.sismember(quoted_set_name, quoted_member)

    def cache_guess_similarity(self, word_id: int, guess: str, similarity: float, ttl: int = 86400):
        """
        缓存猜词相似度
        Args:
            word_id: 关卡/目标词ID
            guess: 猜测词
            similarity: 相似度值
            ttl: 过期时间（秒），默认24小时
        """
        if not self.is_connected():
            return

        try:
            cache_key = f"/wordGame/{word_id}/guesses"
            similarity_key = f"/wordGame/{word_id}/similarity/{guess}"

            self.client.sadd(cache_key, guess)
            self.client.setex(similarity_key, ttl, str(similarity))
            self.client.expire(cache_key, ttl)
        except Exception as e:
            print(f"Redis缓存写入失败: {e}")

    def get_cached_similarity(self, word_id: int, guess: str) -> Optional[float]:
        """
        获取缓存的猜词相似度
        Args:
            word_id: 关卡/目标词ID
            guess: 猜测词
        Returns:
            缓存的相似度值，如果没有缓存则返回None
        """
        if not self.is_connected():
            return None

        try:
            cache_key = f"/wordGame/{word_id}/guesses"
            if not self.client.sismember(cache_key, guess):
                return None

            similarity_key = f"/wordGame/{word_id}/similarity/{guess}"
            similarity_str = self.client.get(similarity_key)
            if similarity_str:
                return float(similarity_str)
        except Exception as e:
            print(f"Redis缓存读取失败: {e}")
        return None

    def get_cached_guesses(self, word_id: int) -> list:
        """
        获取某关卡所有缓存的猜测词
        Args:
            word_id: 关卡/目标词ID
        Returns:
            猜测词列表
        """
        if not self.is_connected():
            return []

        try:
            cache_key = f"/wordGame/{word_id}/guesses"
            return list(self.client.smembers(cache_key))
        except Exception as e:
            print(f"Redis缓存读取失败: {e}")
        return []

    def clear_guess_cache(self, word_id: int):
        """
        清除某关卡的猜词缓存
        Args:
            word_id: 关卡/目标词ID
        """
        if not self.is_connected():
            return

        try:
            cache_key = f"/wordGame/{word_id}/guesses"
            guesses = self.client.smembers(cache_key)

            for guess in guesses:
                similarity_key = f"/wordGame/{word_id}/similarity/{guess}"
                self.client.delete(similarity_key)

            self.client.delete(cache_key)
        except Exception as e:
            print(f"Redis缓存清除失败: {e}")

    def cache_guess_words_list(self, words_data: str, ttl: int = 10800):
        """
        缓存猜字列表
        Args:
            words_data: JSON序列化的猜字列表数据
            ttl: 过期时间（秒），默认3小时
        """
        if not self.is_connected():
            return

        try:
            cache_key = "/wordGame/guess_words/list"
            self.client.setex(cache_key, ttl, words_data)
        except Exception as e:
            print(f"Redis缓存写入失败: {e}")

    def get_cached_guess_words_list(self) -> Optional[str]:
        """
        获取缓存的猜字列表
        Returns:
            JSON序列化的猜字列表数据，如果没有缓存则返回None
        """
        if not self.is_connected():
            return None

        try:
            cache_key = "/wordGame/guess_words/list"
            return self.client.get(cache_key)
        except Exception as e:
            print(f"Redis缓存读取失败: {e}")
        return None

    def invalidate_guess_words_cache(self):
        """
        使猜字列表缓存失效
        """
        if not self.is_connected():
            return

        try:
            cache_key = "/wordGame/guess_words/list"
            self.client.delete(cache_key)
        except Exception as e:
            print(f"Redis缓存清除失败: {e}")


# 全局Redis服务实例
redis_service = RedisService()

File: blog-agent/agent/src/main/java/com/xiafan/agent/service/RedisService.java
package com.xiafan.agent.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.xiafan.agent.config.AppProperties;
import jakarta.annotation.PreDestroy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;

/** Mirrors redisService.py (sessions, identity-set membership check, guess-word caches). */
@Service
public class RedisService {

    private static final Logger log = LoggerFactory.getLogger(RedisService.class);
    private static final String GUESS_WORDS_LIST_KEY = "/wordGame/guess_words/list";

    private final AppProperties props;
    private final ObjectMapper om;
    private volatile JedisPool pool;

    public RedisService(AppProperties props, ObjectMapper om) {
        this.props = props;
        this.om = om;
        tryConnect();
    }

    private void tryConnect() {
        try {
            AppProperties.RedisConfig c = props.getRedis();
            JedisPoolConfig cfg = new JedisPoolConfig();
            cfg.setMaxTotal(16);
            cfg.setTestOnBorrow(true);
            String password = (c.getPassword() == null || c.getPassword().isEmpty()) ? null : c.getPassword();
            pool = new JedisPool(cfg, c.getHost(), c.getPort(), 5000, password, c.getDatabase());
            try (Jedis j = pool.getResource()) {
                j.ping();
            }
            log.info("Redis connected at {}:{}", c.getHost(), c.getPort());
        } catch (Exception e) {
            log.warn("Redis connection failed: {}", e.getMessage());
            pool = null;
        }
    }

    public boolean isConnected() {
        if (pool == null) {
            return false;
        }
        try (Jedis j = pool.getResource()) {
            j.ping();
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    @PreDestroy
    public void close() {
        if (pool != null) {
            pool.close();
        }
    }

    public Optional<String> createSession(int userId, Map<String, Object> userData) {
        if (!isConnected()) {
            return Optional.empty();
        }
        try {
            String sessionId = UUID.randomUUID().toString();
            String sessionKey = "session:" + sessionId;
            String userKey = "user_session:" + userId;
            Map<String, Object> sessionData = new java.util.HashMap<>();
            sessionData.put("user_id", userId);
            sessionData.put("created_at", LocalDateTime.now().toString());
            if (userData != null) {
                sessionData.putAll(userData);
            }
            try (Jedis j = pool.getResource()) {
                j.setex(sessionKey, props.getRedis().getSessionTtl(), om.writeValueAsString(sessionData));
                j.sadd(userKey, sessionId);
                j.expire(userKey, props.getRedis().getSessionTtl());
            }
            return Optional.of(sessionId);
        } catch (Exception e) {
            log.warn("create session failed: {}", e.getMessage());
            return Optional.empty();
        }
    }

    public Optional<Map<String, Object>> getSession(String sessionId) {
        if (!isConnected()) {
            return Optional.empty();
        }
        try {
            String sessionKey = "session:" + sessionId;
            try (Jedis j = pool.getResource()) {
                String data = j.get(sessionKey);
                if (data == null) {
                    return Optional.empty();
                }
                j.expire(sessionKey, props.getRedis().getSessionTtl());
                return Optional.of(om.readValue(data, new TypeReference<Map<String, Object>>() {
                }));
            }
        } catch (Exception e) {
            log.warn("get session failed: {}", e.getMessage());
            return Optional.empty();
        }
    }

    public boolean deleteSession(String sessionId) {
        if (!isConnected()) {
            return false;
        }
        try {
            String sessionKey = "session:" + sessionId;
            try (Jedis j = pool.getResource()) {
                String data = j.get(sessionKey);
                if (data != null) {
                    Map<String, Object> sessionData = om.readValue(data, new TypeReference<Map<String, Object>>() {
                    });
                    Object userId = sessionData.get("user_id");
                    j.del(sessionKey);
                    if (userId != null) {
                        j.srem("user_session:" + userId, sessionId);
                    }
                }
            }
            return true;
        } catch (Exception e) {
            log.warn("delete session failed: {}", e.getMessage());
            return false;
        }
    }

    public boolean deleteUserSessions(int userId) {
        if (!isConnected()) {
            return false;
        }
        try {
            String userKey = "user_session:" + userId;
            try (Jedis j = pool.getResource()) {
                Set<String> sessionIds = j.smembers(userKey);
                for (String sessionId : sessionIds) {
                    j.del("session:" + sessionId);
                }
                j.del(userKey);
            }
            return true;
        } catch (Exception e) {
            log.warn("delete user sessions failed: {}", e.getMessage());
            return false;
        }
    }

    public boolean check(String token) {
        if (pool == null) {
            return false;
        }
        try (Jedis j = pool.getResource()) {
            return j.exists("token:" + token);
        } catch (Exception e) {
            return false;
        }
    }

    public boolean checkSetMember(String setName, String member) {
        if (pool == null) {
            return false;
        }
        try (Jedis j = pool.getResource()) {
            if (j.sismember(setName, member)) {
                return true;
            }
            String quotedMember = "\"" + member + "\"";
            String quotedSetName = "\"" + setName + "\"";
            return j.sismember(quotedSetName, quotedMember);
        } catch (Exception e) {
            return false;
        }
    }

    public void cacheGuessSimilarity(int wordId, String guess, double similarity, int ttl) {
        if (!isConnected()) {
            return;
        }
        try {
            String cacheKey = "/wordGame/" + wordId + "/guesses";
            String similarityKey = "/wordGame/" + wordId + "/similarity/" + guess;
            try (Jedis j = pool.getResource()) {
                j.sadd(cacheKey, guess);
                j.setex(similarityKey, ttl, String.valueOf(similarity));
                j.expire(cacheKey, ttl);
            }
        } catch (Exception e) {
            log.warn("Redis guess-cache write failed: {}", e.getMessage());
        }
    }

    public Optional<Double> getCachedSimilarity(int wordId, String guess) {
        if (!isConnected()) {
            return Optional.empty();
        }
        try {
            String cacheKey = "/wordGame/" + wordId + "/guesses";
            try (Jedis j = pool.getResource()) {
                if (!j.sismember(cacheKey, guess)) {
                    return Optional.empty();
                }
                String similarityKey = "/wordGame/" + wordId + "/similarity/" + guess;
                String s = j.get(similarityKey);
                if (s != null && !s.isEmpty()) {
                    return Optional.of(Double.parseDouble(s));
                }
            }
        } catch (Exception e) {
            log.warn("Redis guess-cache read failed: {}", e.getMessage());
        }
        return Optional.empty();
    }

    public Set<String> getCachedGuesses(int wordId) {
        if (!isConnected()) {
            return Set.of();
        }
        try {
            String cacheKey = "/wordGame/" + wordId + "/guesses";
            try (Jedis j = pool.getResource()) {
                return new HashSet<>(j.smembers(cacheKey));
            }
        } catch (Exception e) {
            log.warn("Redis guess-cache read failed: {}", e.getMessage());
            return Set.of();
        }
    }

    public void clearGuessCache(int wordId) {
        if (!isConnected()) {
            return;
        }
        try {
            String cacheKey = "/wordGame/" + wordId + "/guesses";
            try (Jedis j = pool.getResource()) {
                Set<String> guesses = j.smembers(cacheKey);
                for (String guess : guesses) {
                    j.del("/wordGame/" + wordId + "/similarity/" + guess);
                }
                j.del(cacheKey);
            }
        } catch (Exception e) {
            log.warn("Redis guess-cache clear failed: {}", e.getMessage());
        }
    }

    public void cacheGuessWordsList(String wordsData, int ttl) {
        if (!isConnected()) {
            return;
        }
        try (Jedis j = pool.getResource()) {
            j.setex(GUESS_WORDS_LIST_KEY, ttl, wordsData);
        } catch (Exception e) {
            log.warn("Redis words-list cache write failed: {}", e.getMessage());
        }
    }

    public Optional<String> getCachedGuessWordsList() {
        if (!isConnected()) {
            return Optional.empty();
        }
        try (Jedis j = pool.getResource()) {
            String v = j.get(GUESS_WORDS_LIST_KEY);
            return v == null ? Optional.empty() : Optional.of(v);
        } catch (Exception e) {
            log.warn("Redis words-list cache read failed: {}", e.getMessage());
            return Optional.empty();
        }
    }

    public void invalidateGuessWordsCache() {
        if (!isConnected()) {
            return;
        }
        try (Jedis j = pool.getResource()) {
            j.del(GUESS_WORDS_LIST_KEY);
        } catch (Exception e) {
            log.warn("Redis words-list cache clear failed: {}", e.getMessage());
        }
    }
}